python/src/aiconfig/editor/server/server.py

```diff
@@ -207,28 +217,62 @@ def run() -> FlaskResponse:
     # Define stream callback and queue object for streaming results
     output_text_queue = QueueIterator()
 
-    def update_output_queue(data, _accumulated_data, _index) -> None:
-        should_end_stream = data == STOP_STREAMING_SIGNAL
+    def update_output_queue(data: str, _accumulated_data: str, _index: int) -> None:
+        should_end_stream: bool = data == STOP_STREAMING_SIGNAL
         output_text_queue.put(data, should_end_stream)
 
     inference_options = InferenceOptions(
         stream=stream,
         stream_callback=update_output_queue,
     )
 
-    def generate():
+    # Deepcopy the aiconfig prior to run so we can restore it in the case the run operation is cancelled or encounters some error
+    aiconfig_deep_copy = copy.deepcopy(aiconfig)
+
+    state.events[cancellation_token_id] = threading.Event()
+
+    def generate(cancellation_token_id: str):  # type: ignore
         # Use multi-threading so that we don't block run command from
         # displaying the streamed output (if streaming is supported)
         def run_async_config_in_thread():
-            asyncio.run(
-                aiconfig.run(
-                    prompt_name=prompt_name,
-                    params=params,
-                    run_with_dependencies=False,
-                    options=inference_options,
-                )
-            )
-            output_text_queue.put(STOP_STREAMING_SIGNAL)
+            asyncio.run(aiconfig.run(prompt_name=prompt_name, params=params, run_with_dependencies=False, options=inference_options))  # type: ignore
+            output_text_queue.put(STOP_STREAMING_SIGNAL)  # type: ignore
+
+        def create_cancellation_payload():
+            aiconfig_json = aiconfig_deep_copy.model_dump(exclude=EXCLUDE_OPTIONS) if aiconfig_deep_copy is not None else None
+            return json.dumps({"error": {"message": "The task was cancelled.", "code": 499, "data": aiconfig_json}})
+
+        def handle_cancellation():
+            yield "["
+            yield create_cancellation_payload()
+            yield "]"
+
+            # Reset the aiconfig state to the state prior to the run, and kill the running thread
+            kill_thread(t.ident)
+            state.aiconfig = aiconfig_deep_copy
+
+        def kill_thread(thread_id: int | None):
+            """
+            Kill the thread with the given thread_id.
+
+            PyThreadState_SetAsyncExc: This is a C API function in Python which is used to raise an exception in the context
+            of the specified thread.
+
+            SystemExit: This is the exception we'd like to raise in the target thread.
+            """
+            if thread_id is None:
+                # Nothing to do
+                return
+
+            response = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_long(thread_id), ctypes.py_object(SystemExit))
```

Comment:
Ideally we shouldn't be this low-level. I think if this works it's ok for now, but we should figure out the best supported way to do this. create_task(asyncio.to_thread()) might be it, might not, TBD

Reply:
Going to leave as-is -- couldn't figure out a great way to terminate a thread in Python, other than this. Can revisit if needed

Reply:
This is fine for now if it works. Whenever we're ready to migrate to V2, I'm pretty sure it'll just work using asyncio to_thread() and cancel, for any operation (not just run).

Reply:
Does v2 have streaming set up? I don't think we should spend any time on v2 until we get critical things done first.